pkg/operator/oidcdiscoveryendpoint/controller.go

```diff
@@ -0,0 +1,413 @@
+			// Be safe and don't react.  When the openshift-kube-apiserver-operator
+			// recreates it, we'll catch it.
+			return false
+		},
+	}
+
+	// Create a namespace local cache separate from the Manager cache
+	// A namespace scoped cache can still handle cluster scoped resources
+	cache, err := cache.New(config, cache.Options{Namespace: tokenSignerConfigMapNamespace})
+	if err != nil {
+		return err
+	}
+
+	informer, err := cache.GetInformer(context.TODO(), &corev1.ConfigMap{})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Informer{Informer: informer}, &handler.EnqueueRequestForObject{}, p)
+	if err != nil {
+		return err
+	}
+
+	mgr.Add(&oidcDiscoveryEndpointController{reconciler: r, cache: cache, logger: logger})
+
+	return nil
+}
+
+// openshift-kube-apiserver/bound-sa-token-signing-certs[service-account-xxx.pub]
+func isServiceAccountTokenSigner(meta metav1.Object) bool {
+	// all managed resources are named pod-identity-webhook
+	if meta.GetName() == tokenSignerConfigMapName {
+		return true
+	}
+	return false
+}
+
+type s3EndpointReconciler struct {
+	kubeclientset      *kubernetes.Clientset
+	configclientset    *configset.Clientset
+	logger             log.FieldLogger
+	eventRecorder      events.Recorder
+	infrastructureName string
+	region             string
+}
+
+var _ reconcile.Reconciler = &s3EndpointReconciler{}
+
+func (r *s3EndpointReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	r.logger.Info("reconciling AWS S3 OIDC discovery endpoint on detected change to service account token signer")
+
+	skip, err := r.reconcileClusterAuthentication()
+	if err != nil {
+		r.logger.Error(err)
+		return reconcile.Result{Requeue: true}, err
+	}
+	if skip {
+		return reconcile.Result{}, nil
+	}
+
+	err = r.reconcileS3Endpoint()
+	if err != nil {
+		r.logger.Error(err)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *s3EndpointReconciler) reconcileClusterAuthentication() (bool, error) {
+	r.logger.Debugf("reconciling cluster Authentication serviceAccountIssuer")
+
+	auth, err := r.configclientset.ConfigV1().Authentications().Get(context.TODO(), "cluster", metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+
+	serviceAccountIssuer := r.getIssuerURL()
+	if auth.Spec.ServiceAccountIssuer != "" {
+		r.logger.Debug("Authentication ServiceAccountIssuer already specified")
+		if auth.Spec.ServiceAccountIssuer != serviceAccountIssuer {
+			r.logger.Debug("Authentication ServiceAccountIssuer is not set to the s3 bucket location, skipping OIDC reconciliation")
+			return true, nil
+		}
+		return false, nil
+	}
+
+	auth.Spec.ServiceAccountIssuer = serviceAccountIssuer
+	_, err = r.configclientset.ConfigV1().Authentications().Update(context.TODO(), auth, metav1.UpdateOptions{})
+	return false, err
+}
+
+func (r *s3EndpointReconciler) getIssuerURL() string {
+	return fmt.Sprintf("https://s3-%s.amazonaws.com/%s", r.region, r.getBucketName())
```

Comment:
Will there be issues here with govcloud or any non-standard AWS endpoints?

Reply:
yes.  i might look at how image-registry handles this. it is the only other cluster component i know of that writes to s3.

Reply:
registry is has a public block by default and doesn't need to know the public URL if it is made public.  so... hmmm.

Reply:
I'm not sure how to interpret the data at https://docs.aws.amazon.com/govcloud-us/latest/UserGuide/using-govcloud-endpoints.html , but it seems to indicate that only `us-gov-west-1` would be valid...

And https://github.com/aws/aws-sdk-go/blob/master/aws/endpoints/defaults.go#L7876-L7891 makes it seem like we can just use the `region` as the code is presently doing.

Reply:
https://github.com/terraform-providers/terraform-provider-aws/blob/934b07ac2dc27a97e041e5f8558d2dd3bf369ae1/aws/resource_aws_s3_bucket.go#L1722-L1734

Reply:
see https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/ the style you are using will be deprecated in Sep 2020